adagrams/game.py

```diff
@@ -1,11 +1,145 @@
+import random
+
 def draw_letters():
-    pass
+    # define the dictionary with the letters and their value
+    LETTER_POOL = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+    }
+
+    def pick_random_letter():
+        # choose a letter randomly from the dict
+        letters = [letter for letter, value in LETTER_POOL.items() if value > 0] # looping through LETTER_POOL to get the value of the letters only if the value is bigger than 0
+        if not letters:
+            return None    #if there are no available letters, return None
+        # use random.randint to pick a letter by index
+        index = random.randint(0, len(letters) - 1)        # generates a random index between 0 and len(letters) - 1
+        return letters[index]
+    
+    draw_letters = []
+
+    for i in range(10):    #select exactly 10 letters from dict
+        letter = pick_random_letter()
+        if letter is None:         # if None is returned indicating no more letters available, break the loop
+            break
+        draw_letters.append(letter)    # adding the randomly picked letters into the initial empty list
+        LETTER_POOL[letter] -= 1       # decrease the count of the drawn letter in the dict
+        if LETTER_POOL[letter] == 0:   # if the count of a letter reaches 0, removes the letter from the dict
+            del LETTER_POOL[letter] # remove the letter if its value reaches 0
+
+    return draw_letters
```

Comment:
There is some really cool stuff in this function/algorithm that you wrote. The nested function is well written and can be useful in a variety of situations. Typically however, we'll want to use a nested function only if it makes sense and I don't think this is one of those situations. Your pick_random_letter() function is technically doing what you want it, but you are making certain checks that we don't actually need to make! In a sense, you are overcomplicating your code here. 

One place where I think there's a bit of confusion is how you are handling LETTER_POOL. It is true that you are destroying LETTER_POOL as you remove the tiles that get added to your hand, so you might potentially need to make checks to see if you have enough tiles left. My concern with this particular nested function is that it recreates the list comprehension on line 36 every time it gets called. It would be much better if we had a more direct way of figuring out our range for randint. 

I think that overall, the best way to refactor this would be to use LETTER_POOL to make a list of every single potential letter. From there, you can run your while loop. Within the while loop, use randint to select single letters from the newly created list. Because that list is essentially a copy, we can remove the letters and essentially destroy the list without worrying about needing to use it again later.